lib/broadway_cloud_pub_sub/producer.ex

```diff
@@ -143,27 +104,60 @@ defmodule BroadwayCloudPubSub.Producer do
 
   use GenStage
   alias Broadway.Producer
-  alias BroadwayCloudPubSub.Acknowledger
+  alias BroadwayCloudPubSub.{Acknowledger, Options}
 
   @behaviour Producer
 
-  @default_base_url "https://pubsub.googleapis.com"
-  @default_client BroadwayCloudPubSub.PullClient
-  @default_receive_interval 5000
+  @impl GenStage
+  def init(opts) do
+    receive_interval = opts[:receive_interval]
+    client = opts[:client]
+
+    {:ok, config} = client.init(opts)
+    ack_ref = opts[:broadway][:name]
+
+    {:producer,
+     %{
+       demand: 0,
+       receive_timer: nil,
+       receive_interval: receive_interval,
+       client: {client, config},
+       ack_ref: ack_ref
+     }}
+  end
 
   @impl Producer
-  def prepare_for_start(module, opts) do
-    {me, my_opts} = opts[:producer][:module]
-    client = Keyword.get(my_opts, :client, @default_client)
+  def prepare_for_start(module, broadway_opts) do
+    {producer_module, client_opts} = broadway_opts[:producer][:module]
 
-    my_opts =
-      Keyword.put_new_lazy(my_opts, :pool_size, fn ->
-        2 * opts[:producer][:concurrency]
+    client_opts =
+      Keyword.put_new_lazy(client_opts, :pool_size, fn ->
+        2 * broadway_opts[:producer][:concurrency]
       end)
 
-    {specs, my_opts} = prepare_to_connect(module, client, my_opts)
+    case NimbleOptions.validate(client_opts, Options.definition()) do
+      {:error, error} ->
+        raise ArgumentError,
+              Options.format_error(error, __MODULE__, :prepare_for_start, 2)
 
-    {specs, put_in(opts, [:producer, :module], {me, my_opts})}
+      {:ok, opts} ->
+        ack_ref = broadway_opts[:name]
+        client = opts[:client]
+
+        opts =
+          Keyword.put_new_lazy(opts, :token_generator, fn ->
+            Options.make_token_generator(opts)
+          end)
+
+        {specs, opts} = prepare_to_connect(module, client, opts)
+
+        :persistent_term.put(ack_ref, Map.new(opts))
```

Comment:
I am not thrilled with this change. On the good side, there is no reason to store a separate term per client - they all use the same configuration. On the bad side, the map in persistent term represents the config passed to the client by Acknowledger, whereas the config returned by the client on init/1 represents the config passed to the client by the Producer.

This dual-config issue is fine for the default client, but it may make it difficult to implement a custom client. Wdyt?

Reply:
It is fine either way, keeping decoupled is not an issue though if we are increasing the number of configs to two. At same time, if someone wants distinct configs for those, they can always provide some nesting inside `opts`? 🤔

Reply:
Works for me! I just wanted to point it out :)